BadHabbit/BadHabbit/UserStories/SignInScreen/SignInViewController.swift

```diff
@@ -0,0 +1,135 @@
+            givenNameLabel.topAnchor.constraint(equalTo: userIdentifierLabel.bottomAnchor, constant: 15.0)
+        ])
+
+        familyNameLabel.translatesAutoresizingMaskIntoConstraints = false
+        view.addSubview(familyNameLabel)
+        NSLayoutConstraint.activate([
+            familyNameLabel.rightAnchor.constraint(equalTo: view.rightAnchor, constant: -15.0),
+            familyNameLabel.leftAnchor.constraint(equalTo: view.leftAnchor, constant: 15.0),
+            familyNameLabel.heightAnchor.constraint(equalToConstant: 50.0),
+            familyNameLabel.topAnchor.constraint(equalTo: givenNameLabel.bottomAnchor, constant: 15.0)
+        ])
+
+        emailLabel.translatesAutoresizingMaskIntoConstraints = false
+        view.addSubview(emailLabel)
+        NSLayoutConstraint.activate([
+            emailLabel.rightAnchor.constraint(equalTo: view.rightAnchor, constant: -15.0),
+            emailLabel.leftAnchor.constraint(equalTo: view.leftAnchor, constant: 15.0),
+            emailLabel.heightAnchor.constraint(equalToConstant: 50.0),
+            emailLabel.topAnchor.constraint(equalTo: familyNameLabel.bottomAnchor, constant: 15.0)
+        ])
+
+    }
+
+    private func showUserData(userIdentifier: String, givenName: String?, familyName: String?, email: String?) {
+        setupUserLables()
+        userIdentifierLabel.text = userIdentifier
+        givenNameLabel.text = givenName
+        familyNameLabel.text = familyName
+        emailLabel.text = email
+    }
+
+    @objc
+    func handleAuthorizationAppleIDButtonPress() {
+        let appleIDProvider = ASAuthorizationAppleIDProvider()
+        let request = appleIDProvider.createRequest()
+        request.requestedScopes = [.fullName, .email]
+
+        let authorizationController = ASAuthorizationController(authorizationRequests: [request])
+        authorizationController.delegate = self
+        authorizationController.presentationContextProvider = self
+        authorizationController.performRequests()
+    }
+}
+
+extension SignInViewController: ASAuthorizationControllerDelegate {
+
+    func authorizationController(controller: ASAuthorizationController, didCompleteWithAuthorization authorization: ASAuthorization) {
+        switch authorization.credential {
+        case let appleIDCredential as ASAuthorizationAppleIDCredential:
+            let userIdentifier = appleIDCredential.user
+            let fullName = appleIDCredential.fullName
+            let email = appleIDCredential.email
+            showUserData(userIdentifier: userIdentifier, givenName: fullName?.givenName, familyName: fullName?.familyName, email: email)
+        case let passwordCredential as ASPasswordCredential:
+            let username = passwordCredential.user
+            let password = passwordCredential.password
+
+            DispatchQueue.main.async {
+                self.showPasswordCredentialAlert(username: username, password: password)
```

Comment:
why this call from main queue, but `showUserData` called earlier not from main?

Reply:
changed